Compilation error (missing test_type) on latest nightlies

As a result of https://github.com/rust-lang/rust/pull/64873, merged 6 days ago, `rustc_test::TestDesc` now needs a `test_type` field. I'm guessing that will be `TestType:: IntegrationTest`.

```
error[E0063]: missing field `test_type` in initializer of `rustc_test::TestDesc`
   --> /Users/.../.cargo/registry/src/github.com-1ecc6299db9ec823/datatest-0.5.2/src/runner.rs:171:23
    |
171 |                 desc: TestDesc {
    |                       ^^^^^^^^ missing `test_type`
```

I don't know how this will interact with the whole 'working on stable Rust' goal.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -22,6 +22,7 @@ serde = "1.0.84"
 serde_yaml = "0.8.7"
 yaml-rust = "0.4.2"
 ctor = "0.1.10"
+cfg-if = "0.1.10"
 region = { version = "2.1.2", optional = true }
 
 [dev-dependencies]

--- a/azure-pipelines.yaml
+++ b/azure-pipelines.yaml
@@ -8,7 +8,7 @@ jobs:
   - template: ci/job-rustfmt.yml
   - template: ci/job-check.yml
     parameters:
-      toolchain: nightly-2019-09-13
+      toolchain: nightly-2019-10-19
   - template: ci/job-test.yml
     parameters:
       name: test_stable
@@ -21,4 +21,4 @@ jobs:
   - template: ci/job-test.yml
     parameters:
       name: test_nightly
-      toolchain: nightly-2019-09-13
+      toolchain: nightly-2019-10-19

--- a/azure-pipelines.yaml
+++ b/azure-pipelines.yaml
@@ -8,7 +8,7 @@ jobs:
   - template: ci/job-rustfmt.yml
   - template: ci/job-check.yml
     parameters:
-      toolchain: nightly-2019-10-19
+      toolchain: nightly-2019-11-02
   - template: ci/job-test.yml
     parameters:
       name: test_stable
@@ -21,4 +21,4 @@ jobs:
   - template: ci/job-test.yml
     parameters:
       name: test_nightly
-      toolchain: nightly-2019-10-19
+      toolchain: nightly-2019-11-02

--- a/azure-pipelines.yaml
+++ b/azure-pipelines.yaml
@@ -9,6 +9,7 @@ jobs:
   - template: ci/job-check.yml
     parameters:
       toolchain: nightly-2019-11-02
+      features: ["post_v139"]
   - template: ci/job-test.yml
     parameters:
       name: test_stable
@@ -22,3 +23,4 @@ jobs:
     parameters:
       name: test_nightly
       toolchain: nightly-2019-11-02
+      features: ["post_v139"]

--- a/ci/job-check.yml
+++ b/ci/job-check.yml
@@ -1,5 +1,6 @@
 parameters:
   toolchain: stable
+  features: []
 jobs:
   - job: check
     pool:
@@ -11,7 +12,7 @@ jobs:
           components:
             - clippy
       - script: |
-          cargo check --all --all-targets
+          cargo check --all --all-targets --features "${{ join(' ', parameters.features) }}"
         displayName: cargo check
       - script: |
           cargo clippy --all

--- a/ci/job-check.yml
+++ b/ci/job-check.yml
@@ -15,6 +15,6 @@ jobs:
           cargo check --all --all-targets --features "${{ join(' ', parameters.features) }}"
         displayName: cargo check
       - script: |
-          cargo clippy --all
+          cargo clippy --all --features "${{ join(' ', parameters.features) }}"
         displayName: cargo clippy --all
 

--- a/datatest-derive/src/lib.rs
+++ b/datatest-derive/src/lib.rs
@@ -284,6 +284,7 @@ fn files_internal(
             pattern: #pattern_idx,
             ignorefn: #ignore_func_ref,
             testfn: ::datatest::__internal::FilesTestFn::#kind(#trampoline_func_ident),
+            source_file: file!(),
         };
 
         #[automatically_derived]
@@ -523,6 +524,7 @@ fn data_internal(
             name: concat!(module_path!(), "::", #func_name_str),
             ignore: #ignore,
             describefn: #describe_func_ident,
+            source_file: file!(),
         };
 
         #[automatically_derived]
@@ -631,6 +633,7 @@ pub fn test_ctor_registration(
                 ::datatest::__internal::assert_test_result(result);
             },
             should_panic: #should_panic,
+            source_file: file!(),
         };
 
         #func_item

--- a/src/data.rs
+++ b/src/data.rs
@@ -11,6 +11,7 @@ pub struct DataTestDesc {
     pub name: &'static str,
     pub ignore: bool,
     pub describefn: fn() -> Vec<DataTestCaseDesc<DataTestFn>>,
+    pub source_file: &'static str,
 }
 
 /// Used internally for `#[datatest::data(..)]` tests.

--- a/src/files.rs
+++ b/src/files.rs
@@ -21,6 +21,7 @@ pub struct FilesTestDesc {
     pub pattern: usize,
     pub ignorefn: Option<fn(&Path) -> bool>,
     pub testfn: FilesTestFn,
+    pub source_file: &'static str,
 }
 
 /// Trait defining conversion into a function argument. We use it to convert discovered paths

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -199,3 +199,20 @@ fn read_to_end(path: &Path) -> Vec<u8> {
         .unwrap_or_else(|e| panic!("cannot read test input at '{}': {}", path.display(), e));
     input
 }
+
+use crate::rustc_test::TestType;
+
+/// Helper function used internally, to mirror how rustc_test chooses a TestType.
+/// Must be called with the result of `file!()` (called in macro output) to be meaningful.
+pub fn test_type(path: &'static str) -> TestType {
+    if path.starts_with("src") {
+        // `/src` folder contains unit-tests.
+        TestType::UnitTest
+    } else if path.starts_with("tests") {
+        // `/tests` folder contains integration tests.
+        TestType::IntegrationTest
+    } else {
+        // Crate layout doesn't match expected one, test type is unknown.
+        TestType::Unknown
+    }
+}

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -200,10 +200,12 @@ fn read_to_end(path: &Path) -> Vec<u8> {
     input
 }
 
+#[cfg(feature = "post_v139")]
 use crate::rustc_test::TestType;
 
 /// Helper function used internally, to mirror how rustc_test chooses a TestType.
 /// Must be called with the result of `file!()` (called in macro output) to be meaningful.
+#[cfg(feature = "post_v139")]
 pub fn test_type(path: &'static str) -> TestType {
     if path.starts_with("src") {
         // `/src` folder contains unit-tests.

--- a/src/runner.rs
+++ b/src/runner.rs
@@ -49,23 +49,6 @@ fn derive_test_name(root: &Path, path: &Path, test_name: &str) -> String {
     test_name
 }
 
-fn test_desc(
-    name: TestName,
-    ignore: bool,
-    should_panic: ShouldPanic,
-    allow_fail: bool,
-    source_file: &'static str,
-) -> TestDesc {
-    TestDesc {
-        name,
-        ignore,
-        should_panic,
-        allow_fail,
-        #[cfg(feature = "post_v139")]
-        test_type: crate::test_type(source_file),
-    }
-}
-
 /// When compiling tests, Rust compiler collects all items marked with `#[test_case]` and passes
 /// references to them to the test runner in a slice (like `&[&test_a, &test_b, &test_c]`). Since
 /// we need a different descriptor for our data-driven tests than the standard one, we have two
@@ -186,14 +169,15 @@ fn render_files_test(desc: &FilesTestDesc, rendered: &mut Vec<TestDescAndFn>) {
 
             // Generate a standard test descriptor
             let desc = TestDescAndFn {
-                desc: test_desc(
-                    TestName::DynTestName(test_name),
+                desc: TestDesc {
+                    name: TestName::DynTestName(test_name),
                     ignore,
-                    ShouldPanic::No,
+                    should_panic: ShouldPanic::No,
                     // Cannot be used on stable: https://github.com/rust-lang/rust/issues/46488
-                    false,
-                    desc.source_file,
-                ),
+                    allow_fail: false,
+                    #[cfg(feature = "post_v139")]
+                    test_type: crate::test_type(desc.source_file),
+                },
                 testfn,
             };
 
@@ -236,6 +220,7 @@ fn render_data_test(desc: &DataTestDesc, rendered: &mut Vec<TestDescAndFn>) {
                 ignore: desc.ignore,
                 should_panic: ShouldPanic::No,
                 allow_fail: false,
+                #[cfg(feature = "post_v139")]
                 test_type: crate::test_type(desc.source_file),
             },
             testfn,
@@ -397,14 +382,15 @@ fn render_test_descriptor(
         }
         DatatestTestDesc::RegularTest(desc) => {
             rendered.push(TestDescAndFn {
-                desc: test_desc(
-                    TestName::StaticTestName(real_name(desc.name)),
-                    desc.ignore,
+                desc: TestDesc {
+                    name: TestName::StaticTestName(real_name(desc.name)),
+                    ignore: desc.ignore,
+                    should_panic: desc.should_panic.into(),
                     // FIXME: should support!
-                    desc.should_panic.into(),
-                    false,
-                    desc.source_file,
-                ),
+                    allow_fail: false,
+                    #[cfg(feature = "post_v139")]
+                    test_type: crate::test_type(desc.source_file),
+                },
                 testfn: TestFn::StaticTestFn(desc.testfn),
             })
         }

--- a/src/runner.rs
+++ b/src/runner.rs
@@ -1,8 +1,6 @@
 use crate::data::{DataTestDesc, DataTestFn};
 use crate::files::{FilesTestDesc, FilesTestFn};
-use crate::rustc_test::{
-    Bencher, ShouldPanic, TestDesc, TestDescAndFn, TestFn, TestName, TestType,
-};
+use crate::rustc_test::{Bencher, ShouldPanic, TestDesc, TestDescAndFn, TestFn, TestName};
 use std::path::{Path, PathBuf};
 use std::sync::atomic::{AtomicBool, AtomicPtr, Ordering};
 
@@ -34,6 +32,7 @@ pub struct RegularTestDesc {
     pub ignore: bool,
     pub testfn: fn(),
     pub should_panic: RegularShouldPanic,
+    pub source_file: &'static str,
 }
 
 fn derive_test_name(root: &Path, path: &Path, test_name: &str) -> String {
@@ -176,7 +175,7 @@ fn render_files_test(desc: &FilesTestDesc, rendered: &mut Vec<TestDescAndFn>) {
                     should_panic: ShouldPanic::No,
                     // Cannot be used on stable: https://github.com/rust-lang/rust/issues/46488
                     allow_fail: false,
-                    test_type: TestType::IntegrationTest,
+                    test_type: crate::test_type(desc.source_file),
                 },
                 testfn,
             };
@@ -220,7 +219,7 @@ fn render_data_test(desc: &DataTestDesc, rendered: &mut Vec<TestDescAndFn>) {
                 ignore: desc.ignore,
                 should_panic: ShouldPanic::No,
                 allow_fail: false,
-                test_type: TestType::IntegrationTest,
+                test_type: crate::test_type(desc.source_file),
             },
             testfn,
         };
@@ -380,7 +379,7 @@ fn render_test_descriptor(
                     // FIXME: should support!
                     should_panic: desc.should_panic.into(),
                     allow_fail: false,
-                    test_type: TestType::IntegrationTest,
+                    test_type: crate::test_type(desc.source_file),
                 },
                 testfn: TestFn::StaticTestFn(desc.testfn),
             })

--- a/src/runner.rs
+++ b/src/runner.rs
@@ -1,7 +1,5 @@
 use crate::data::{DataTestDesc, DataTestFn};
 use crate::files::{FilesTestDesc, FilesTestFn};
-#[cfg(feature = "post_v139")]
-use crate::rustc_test::TestType;
 use crate::rustc_test::{Bencher, ShouldPanic, TestDesc, TestDescAndFn, TestFn, TestName};
 use std::path::{Path, PathBuf};
 use std::sync::atomic::{AtomicBool, AtomicPtr, Ordering};
@@ -57,13 +55,14 @@ fn test_desc(
     ignore: bool,
     should_panic: ShouldPanic,
     allow_fail: bool,
+    source_file: &'static str,
 ) -> TestDesc {
     TestDesc {
         name,
         ignore,
         should_panic,
         allow_fail,
-        test_type: TestType::Unknown,
+        test_type: crate::test_type(source_file),
     }
 }
 
@@ -73,6 +72,7 @@ fn test_desc(
     ignore: bool,
     should_panic: ShouldPanic,
     allow_fail: bool,
+    _source_file: &'static str,
 ) -> TestDesc {
     TestDesc {
         name,
@@ -207,9 +207,9 @@ fn render_files_test(desc: &FilesTestDesc, rendered: &mut Vec<TestDescAndFn>) {
                     ignore,
                     ShouldPanic::No,
                     // Cannot be used on stable: https://github.com/rust-lang/rust/issues/46488
-                    allow_fail: false,
-                    test_type: crate::test_type(desc.source_file),
-                },
+                    false,
+                    desc.source_file,
+                ),
                 testfn,
             };
 
@@ -354,7 +354,7 @@ pub fn register(new: &mut RegistrationNode) {
 #[doc(hidden)]
 pub fn runner(tests: &[&dyn TestDescriptor]) {
     let args = std::env::args().collect::<Vec<_>>();
-    let mut opts = match crate::rustc_test::parse_opts(&args) {
+    let mut opts = match crate::rustc_test::test::parse_opts(&args) {
         Some(Ok(o)) => o,
         Some(Err(msg)) => panic!("{:?}", msg),
         None => return,
@@ -410,10 +410,10 @@ fn render_test_descriptor(
                     TestName::StaticTestName(real_name(desc.name)),
                     desc.ignore,
                     // FIXME: should support!
-                    should_panic: desc.should_panic.into(),
-                    allow_fail: false,
-                    test_type: crate::test_type(desc.source_file),
-                },
+                    desc.should_panic.into(),
+                    false,
+                    desc.source_file,
+                ),
                 testfn: TestFn::StaticTestFn(desc.testfn),
             })
         }

--- a/src/runner.rs
+++ b/src/runner.rs
@@ -49,7 +49,6 @@ fn derive_test_name(root: &Path, path: &Path, test_name: &str) -> String {
     test_name
 }
 
-#[cfg(feature = "post_v139")]
 fn test_desc(
     name: TestName,
     ignore: bool,
@@ -62,23 +61,8 @@ fn test_desc(
         ignore,
         should_panic,
         allow_fail,
+        #[cfg(feature = "post_v139")]
         test_type: crate::test_type(source_file),
-    }
-}
-
-#[cfg(not(feature = "post_v139"))]
-fn test_desc(
-    name: TestName,
-    ignore: bool,
-    should_panic: ShouldPanic,
-    allow_fail: bool,
-    _source_file: &'static str,
-) -> TestDesc {
-    TestDesc {
-        name,
-        ignore,
-        should_panic,
-        allow_fail,
     }
 }
 
@@ -354,7 +338,14 @@ pub fn register(new: &mut RegistrationNode) {
 #[doc(hidden)]
 pub fn runner(tests: &[&dyn TestDescriptor]) {
     let args = std::env::args().collect::<Vec<_>>();
-    let mut opts = match crate::rustc_test::test::parse_opts(&args) {
+    cfg_if::cfg_if! {
+        if #[cfg(feature = "post_v139")] {
+            let parsed = crate::rustc_test::test::parse_opts(&args);
+        } else {
+            let parsed = crate::rustc_test::parse_opts(&args);
+        }
+    };
+    let mut opts = match parsed {
         Some(Ok(o)) => o,
         Some(Err(msg)) => panic!("{:?}", msg),
         None => return,

--- a/src/runner.rs
+++ b/src/runner.rs
@@ -1,6 +1,8 @@
 use crate::data::{DataTestDesc, DataTestFn};
 use crate::files::{FilesTestDesc, FilesTestFn};
-use crate::rustc_test::{Bencher, ShouldPanic, TestDesc, TestDescAndFn, TestFn, TestName};
+use crate::rustc_test::{
+    Bencher, ShouldPanic, TestDesc, TestDescAndFn, TestFn, TestName, TestType,
+};
 use std::path::{Path, PathBuf};
 use std::sync::atomic::{AtomicBool, AtomicPtr, Ordering};
 
@@ -174,6 +176,7 @@ fn render_files_test(desc: &FilesTestDesc, rendered: &mut Vec<TestDescAndFn>) {
                     should_panic: ShouldPanic::No,
                     // Cannot be used on stable: https://github.com/rust-lang/rust/issues/46488
                     allow_fail: false,
+                    test_type: TestType::IntegrationTest,
                 },
                 testfn,
             };
@@ -217,6 +220,7 @@ fn render_data_test(desc: &DataTestDesc, rendered: &mut Vec<TestDescAndFn>) {
                 ignore: desc.ignore,
                 should_panic: ShouldPanic::No,
                 allow_fail: false,
+                test_type: TestType::IntegrationTest,
             },
             testfn,
         };
@@ -376,6 +380,7 @@ fn render_test_descriptor(
                     // FIXME: should support!
                     should_panic: desc.should_panic.into(),
                     allow_fail: false,
+                    test_type: TestType::IntegrationTest,
                 },
                 testfn: TestFn::StaticTestFn(desc.testfn),
             })

